source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py

```diff
@@ -347,61 +375,137 @@ def __call__(
-            reset_model: Initialize the model even if it already exists. Defaults to False.
+            model_name: The name of the Theia transformer model to prepare.
+            model_device: The device to store and infer the model on.
 
         Returns:
-            torch.Tensor: the image features, on the same device as the image
+            A dictionary containing the model, preprocess and inference functions.
         """
-        if model_zoo_cfg is not None:  # use other than default
-            self.model_zoo_cfg.update(model_zoo_cfg)
+        from transformers import AutoModel
 
-        if model_name not in self.model_zoo or reset_model:
-            # The following allows to only load a desired subset of a model zoo into GPU memory
-            # as it becomes needed, in a "lazy" evaluation.
-            print(f"[INFO]: Adding {model_name} to the model zoo")
-            self.model_zoo[model_name] = self.model_zoo_cfg[model_name]["model"]()
+        def _load_model() -> torch.nn.Module:
+            """Load the Theia transformer model."""
+            model = AutoModel.from_pretrained(f"theaiinstitute/{model_name}", trust_remote_code=True).eval()
+            return model.to(model_device)
 
-        if model_device is not None and self.model_zoo[model_name].device != model_device:
-            # want to offload vision model inference to another device
-            self.model_zoo[model_name] = self.model_zoo[model_name].to(model_device)
+        def _preprocess_image(img: torch.Tensor) -> torch.Tensor:
+            """Preprocess the image for the Theia transformer model.
 
-        images = image(
-            env=env,
-            sensor_cfg=sensor_cfg,
-            data_type=data_type,
-            convert_perspective_to_orthogonal=convert_perspective_to_orthogonal,
-            normalize=True,  # want this for training stability
-        )
+            Args:
+                img: The image tensor to preprocess. Shape is (num_envs, height, width, channel).
+
+            Returns:
+                The preprocessed image tensor. Shape is (num_envs, height, width, channel).
+            """
+            # Move the image to the model device
+            img = img.to(model_device)
+            # Normalize the image
+            min_img = torch.amin(img, dim=(1, 2), keepdim=True)
+            max_img = torch.amax(img, dim=(1, 2), keepdim=True)
+
+            return (img - min_img) / (max_img - min_img)
+
+        def _inference(model, images: torch.Tensor) -> torch.Tensor:
+            """Inference the Theia transformer model.
+
+            Args:
+                model: The Theia transformer model.
+                images: The preprocessed image tensor. Shape is (num_envs, height, width, channel).
+
+            Returns:
+                The extracted features tensor. Shape is (num_envs, feature_dim).
+            """
+            return model.forward_feature(images, do_rescale=False, interpolate_pos_encoding=True)
```

Comment:
This converts from GPU to CPU when it's not needed: https://github.com/isaac-sim/IsaacLab/pull/1323

                "preprocess": lambda img: {
                    "pixel_values": (  # This deconstructs default inferencing (CPU in transformers) into GPU only operations
                        (  # as advised by first author of Theia paper
                            # Shifts mean, standard deviation, and format to match training set
                            img.permute(
                                0, 3, 1, 2
                            ).float()  # Convert [batch, height, width, 3] -> [batch, 3, height, width]
                            - torch.tensor([0.485, 0.456, 0.406], device=img.device).view(1, 3, 1, 1)
                        )
                        / torch.tensor([0.229, 0.224, 0.225], device=img.device).view(1, 3, 1, 1)
                    ) / 255
                },
                "inference": lambda model, images: model.backbone.model(  # Taken from Transformers; inference converted to be GPU only
                    **images, interpolate_pos_encoding=True
                ).last_hidden_state[
                    :, 1:
                ],

Reply:
Also includes the divide by 255 fix to normalize between 0 and 1